File: src/lib/browser.ts
import path from 'path';
import os from 'os';
import fs from 'fs';
import { chromium, type Browser, type BrowserContext, type Page, type ElementHandle } from 'playwright';
import type { AskOptions } from '../types.js';
import { executeTool } from './toolExecutor.js';

const PLEX_DIR    = path.join(os.homedir(), '.plexcode');
const SESSION_FILE = path.join(PLEX_DIR, 'session.json');
const BASE_URL     = 'https://www.perplexity.ai';

/** Maximum agentic tool-call iterations per user query. */
const MAX_TOOL_LOOPS = 4;

const INPUT_SELECTORS = [
  "[role='textbox']",
  'textarea#ask-input',
  'textarea',
  "[contenteditable='true']",
];

const MODE_MAP: Record<string, string> = {
  'deep-research': 'Deep research',
  'model-council': 'Model council',
  'create':        'Create files and apps',
  'learn':         'Learn step by step',
};

/** System preamble injected into the first message of each chat thread. */
const AGENT_PREAMBLE = `\
[System] You have access to the user's local file system through special tool commands.
To use a tool, reply with an XML-style tag and STOP writing immediately after the closing tag:
<TOOL>command here</TOOL>

Allowed commands: cat, ls, head, tail, grep, find, wc, pwd, file, tree, echo, stat

Examples:
  <TOOL>ls -la src/</TOOL>
  <TOOL>cat src/lib/browser.ts</TOOL>
  <TOOL>grep -rn "TODO" src/</TOOL>
  <TOOL>head -n 50 src/app.tsx</TOOL>

Rules:
1. You may call at most ${MAX_TOOL_LOOPS} tools per question.
2. After emitting a <TOOL> tag, STOP — write nothing else until you receive the result.
3. Once you have enough context, write your final answer as normal markdown WITHOUT any <TOOL> tags.
4. If you don't need file context to answer the question, answer directly without using any tools.

`;

/** Strip any residual <TOOL>...</TOOL> blocks from the final response. */
function cleanResponse(text: string): string {
  return text.replace(/<TOOL>[\s\S]*?<\/TOOL>/gi, '').trim();
}

export class PerplexityBrowser {
  private browser: Browser | null = null;
  private context: BrowserContext | null = null;
  private page: Page | null = null;
  private _interrupted     = false;
  /** Tracks whether the agentic preamble has been injected into the current thread. */
  private _preambleInjected = false;

  hasSession(): boolean {
    return fs.existsSync(SESSION_FILE);
  }

  async start(headless?: boolean): Promise<void> {
    fs.mkdirSync(PLEX_DIR, { recursive: true });
    const runHeadless = headless ?? this.hasSession();

    const launchOpts = {
      headless: runHeadless,
      args: ['--no-sandbox', '--disable-dev-shm-usage', '--disable-blink-features=AutomationControlled'],
    };

    try {
      // Trying local Chrome first avoids Cloudflare blocking automated browsers
      this.browser = await chromium.launch({ ...launchOpts, channel: 'chrome' });
    } catch {
      // Fallback to bundled chromium
      this.browser = await chromium.launch(launchOpts);
    }

    const ctxOpts: Record<string, unknown> = {
      viewport: { width: 1280, height: 900 },
      locale: 'en-US',
      userAgent:
        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36',
    };

    if (fs.existsSync(SESSION_FILE)) {
      ctxOpts['storageState'] = SESSION_FILE;
    }

    this.context = await this.browser.newContext(ctxOpts);
    this.page    = await this.context.newPage();
    await this.page.goto(BASE_URL, { waitUntil: 'domcontentloaded' });
    await this.page.waitForTimeout(2000);
  }

  /** Call after user manually logs in — saves session for next headless run. */
  async saveSession(): Promise<void> {
    await this.context!.storageState({ path: SESSION_FILE });
  }

  async newChat(): Promise<void> {
    if (!this.page) return;
    // Reset preamble so it re-injects on the next message of the new thread.
    this._preambleInjected = false;
    await this.page.goto(BASE_URL, { waitUntil: 'domcontentloaded' });
    await this.page.waitForTimeout(2000);
  }

  interrupt(): void {
    this._interrupted = true;
  }

  // ── Public API ─────────────────────────────────────────────────────────────

  /**
   * Send a prompt to Perplexity and return the final answer string.
   *
   * Implements the Shadow Agentic Loop:
   *  1. Prepend the tool-use system preamble on the first message of the thread.
   *  2. Wait for the AI's response.
   *  3. If the response contains a <TOOL> tag, execute the command locally.
   *  4. Feed the tool output back to Perplexity as a new message.
   *  5. Repeat up to MAX_TOOL_LOOPS times, then return the final answer.
   */
  async ask(prompt: string, opts: AskOptions = {}): Promise<string> {
    const page = this.page!;
    this._interrupted = false;

    // ── 1. Build the augmented prompt (inject preamble once per thread) ──────
    const augmented = this._preambleInjected
      ? prompt
      : AGENT_PREAMBLE + prompt;
    this._preambleInjected = true;

    if (opts.mode && opts.mode !== 'default') {
      await this._selectMode(page, opts.mode);
    }
    if (opts.model) {
      await this._selectModel(page, opts.model);
    }

    // ── 2. Submit the initial user message ───────────────────────────────────
    await this._submitMessage(page, augmented);

    // ── 3. Agentic loop ──────────────────────────────────────────────────────
    let loopCount = 0;

    while (loopCount < MAX_TOOL_LOOPS) {
      await this._waitForDone(page);

      const response = await this._extractResponse(page);

      // Check for a tool call — match the FIRST tag only per response
      const toolMatch = response.match(/<TOOL>([\s\S]*?)<\/TOOL>/i);
      if (!toolMatch) {
        // No tool call → the AI has its final answer
        break;
      }

      loopCount++;
      const command = toolMatch[1].trim();

      // Notify the UI (so the spinner can show live status)
      opts.onAgentLoop?.(loopCount, command);

      // ── Execute the tool locally ─────────────────────────────────────────
      const toolOutput = await executeTool(command);

      // ── Feed result back to Perplexity ───────────────────────────────────
      const feedback =
        `[Tool Result — "${command}"]\n\n${toolOutput}`;

      await this._submitMessage(page, feedback);
    }

    // ── 4. Extract & clean the final answer ──────────────────────────────────
    const finalResponse = await this._extractResponse(page);
    return cleanResponse(finalResponse);
  }

  async stop(): Promise<void> {
    try {
      if (this.context) await this.context.storageState({ path: SESSION_FILE });
      if (this.browser) await this.browser.close();
    } catch { /* ignore */ }
  }

  // ── Private ────────────────────────────────────────────────────────────────

  /**
   * Core message submission — fills the input box and presses Enter.
   * Used by both the initial prompt and autonomous tool-result feedback.
   */
  private async _submitMessage(page: Page, text: string): Promise<void> {
    await page.waitForTimeout(500);

    const input = await this._findInput(page);
    if (!input) throw new Error('Could not find Perplexity input box.');

    await input.click();
    await page.keyboard.press('Control+a');
    await page.keyboard.press('Delete');
    await page.waitForTimeout(300);

    try {
      await input.fill(text);
    } catch {
      // fill() can fail for contenteditable — fall back to chunked typing
      for (let i = 0; i < text.length; i += 1800) {
        await input.type(text.slice(i, i + 1800), { delay: 0 });
      }
    }

    await page.waitForTimeout(500);
    await page.keyboard.press('Enter');
    await page.waitForTimeout(1500);
  }

  private async _findInput(page: Page): Promise<ElementHandle | null> {
    for (const sel of INPUT_SELECTORS) {
      try {
        const el = await page.waitForSelector(sel, { timeout: 6000, state: 'visible' });
        if (el) return el;
      } catch { /* try next */ }
    }
    return null;
  }

  private async _selectMode(page: Page, mode: string): Promise<void> {
    const input = await this._findInput(page);
    if (!input) return;
    await input.click();
    await page.keyboard.type('/');
    await page.waitForTimeout(600);
    const label = MODE_MAP[mode] ?? mode;
    try {
      await page.click(`text=${label}`, { timeout: 3000 });
      await page.waitForTimeout(400);
    } catch {
      await page.keyboard.press('Escape');
    }
  }

  private async _selectModel(page: Page, modelName: string): Promise<void> {
    try {
      await page.click("button[aria-label='Model']", { timeout: 3000 });
      await page.waitForTimeout(500);
      await page.click(`text=${modelName}`, { timeout: 3000 });
      await page.waitForTimeout(300);
    } catch { /* ignore — model stays as is */ }
  }

  private async _waitForDone(page: Page, timeoutMs = 120_000): Promise<void> {
    const stopSel = "button[aria-label*='stop' i], button[aria-label*='cancel' i]";
    const deadline = Date.now() + timeoutMs;
    while (Date.now() < deadline) {
      if (this._interrupted) {
        this._interrupted = false;
        try {
          await page.click(stopSel, { timeout: 1000 });
        } catch { /* ignore if already stopped */ }
        break;
      }

      if ((await page.locator(stopSel).count()) === 0) {
        await page.waitForTimeout(800);
        break;
      }
      await page.waitForTimeout(600);
    }
  }

  private async _extractResponse(page: Page): Promise<string> {
    const selectors = ['.prose', '.markdown', '[data-testid="answer"]', '.cl-prose', 'main'];
    for (const sel of selectors) {
      try {
        const els = await page.$$(sel);
        if (els.length) {
          // Get only the last block to support continuing threads
          const text = await els[els.length - 1].innerText();
          if (text && text.trim()) return text.trim();
        }
      } catch { /* try next */ }
    }
    return '[Could not extract response from Perplexity]';
  }
}

File: browser.py
"""
browser.py  –  Playwright-based Perplexity web session manager.

First run  : opens visible Chromium so you can log in manually.
After login: session is saved to ~/.pplx_proxy_session.json and reused headlessly.
"""
import asyncio, json, os, time
from pathlib import Path
from playwright.async_api import async_playwright, BrowserContext, Page

SESSION_FILE = Path.home() / ".pplx_proxy_session.json"
BASE_URL = "https://www.perplexity.ai"

# CSS selectors – update if Perplexity redesigns their UI
SELECTORS = {
    "textarea":    "textarea",
    "stop_button": "button[aria-label*='stop' i], button[aria-label*='cancel' i]",
    "answer":      [".prose", ".markdown", "[data-testid='answer']", "main"],
}


class PerplexityBrowser:
    def __init__(self, cfg):
        self.cfg = cfg
        self.ready = False
        self._lock = asyncio.Lock()
        self._pw = None
        self._browser = None
        self._ctx: BrowserContext = None
        self._page: Page = None

    # ── Lifecycle ─────────────────────────────────────────────────────────────

    async def start(self):
        self._pw = await async_playwright().start()
        headless = SESSION_FILE.exists()

        self._browser = await self._pw.chromium.launch(
            headless=headless,
            args=["--no-sandbox", "--disable-dev-shm-usage", "--disable-blink-features=AutomationControlled"],
        )

        ctx_opts = {
            "viewport": {"width": 1280, "height": 900},
            "locale": "en-US",
            "user_agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 Chrome/124.0 Safari/537.36"
        }
        if SESSION_FILE.exists():
            ctx_opts["storage_state"] = str(SESSION_FILE)

        self._ctx = await self._browser.new_context(**ctx_opts)
        self._page = await self._ctx.new_page()

        await self._page.goto(BASE_URL, wait_until="domcontentloaded")
        await asyncio.sleep(3)

        if not await self._is_logged_in():
            print("\n[proxy] Browser opened – please log in to Perplexity, then press ENTER here.")
            input()
            await asyncio.sleep(2)

        await self._ctx.storage_state(path=str(SESSION_FILE))
        print("[proxy] Session saved. Future runs will be headless.")
        self.ready = True

    async def stop(self):
        try:
            if self._ctx:
                await self._ctx.storage_state(path=str(SESSION_FILE))
            if self._browser:
                await self._browser.close()
            if self._pw:
                await self._pw.stop()
        except Exception:
            pass

    # ── Core ask ──────────────────────────────────────────────────────────────

    async def ask(self, prompt: str) -> str:
        async with self._lock:
            return await self._do_ask(prompt)

    async def _do_ask(self, prompt: str) -> str:
        page = self._page

        await page.goto(BASE_URL, wait_until="domcontentloaded")
        await asyncio.sleep(1.5)

        try:
            textarea = await page.wait_for_selector(SELECTORS["textarea"], timeout=12_000)
        except Exception:
            return "[proxy error] Could not find Perplexity input box."

        await textarea.click()
        await page.keyboard.press("Control+a")
        await textarea.fill("")

        CHUNK = 1800
        for i in range(0, len(prompt), CHUNK):
            await textarea.type(prompt[i:i+CHUNK], delay=0)

        await page.keyboard.press("Enter")
        await asyncio.sleep(1.5)

        await self._wait_for_done(page)
        return await self._extract(page)

    async def _wait_for_done(self, page: Page, timeout: int = 120):
        """Poll until the stop/cancel button is gone."""
        deadline = time.time() + timeout
        while time.time() < deadline:
            n = await page.locator(SELECTORS["stop_button"]).count()
            if n == 0:
                await asyncio.sleep(0.8)
                break
            await asyncio.sleep(0.6)

    async def _extract(self, page: Page) -> str:
        for sel in SELECTORS["answer"]:
            try:
                els = await page.query_selector_all(sel)
                if els:
                    chunks = [await el.inner_text() for el in els]
                    text = "\n".join(c.strip() for c in chunks if c.strip())
                    if text:
                        return text
            except Exception:
                continue
        return "[proxy error] Could not extract Perplexity response."

    async def _is_logged_in(self) -> bool:
        try:
            el = await self._page.query_selector(SELECTORS["textarea"])
            return el is not None
        except Exception:
            return False
